### ISSUE TITLE
tf.keras.layers.Softmax does not work with axis other than -1

### ISSUE THREAD

**Issue by dd1923** [onnx/keras-onnx#529]
TF 2.2 Keras models having a softmax layer of type tf.keras.layers.Softmax with axis set to anything but -1 fails to produce the correct and comparable results when keras2onnx is used to convert the model to onnx.

Resulting onnx model reports softmax on basically a flatten version of the array instead of doing it specifically along an axis. Note that the axis does get set correctly in the onnx model, however, TF calculates softmax a bit differently than onnx does at the core and hence it results in difference in the output.

TF has two softmax :
tf.keras.layers.Softmax and tf.nn.softmax.

This issue only happens when models use the keras version (which is expected in a majority of keras models) with an axis other than -1. 

Both of the above softmax implementations in TF calls _softmax helper function defined here:
[https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/ops/nn_ops.py#L2968](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/ops/nn_ops.py#L2968)

Notice that is it swapping the axis internally before and after  the softmax. 

Problem does not happen with tf.nn.softmax may be because it is a softmax op in tf namespace and it correctly inserts the transpose ops before and after the softmax (even when keras2onnx is used). I think the reason problem happens in tf.keras.layers.Softmax is because keras2onnx is simply parsing the keras graph, not checking under the hood that the softmax requires additional transpose ops inserted to work correctly.

The above hypothesis also means that it would have worked had it have parsed the full TF graph. I checked it using tf2onnx instead of keras2onnx and it does indeed work. So the bug is in the keras2onnx parser.

Full repro colab link:

[https://colab.research.google.com/drive/11MNXGqnWEaEDWjkhFh9udUDFOqv4pdI4?usp=sharing](https://colab.research.google.com/drive/11MNXGqnWEaEDWjkhFh9udUDFOqv4pdI4?usp=sharing)

The tests are also not covering the tf.keras.layers.Softmax case [https://github.com/onnx/keras-onnx/blob/810cc948fb9c3b0a04345bc556678c03a9b19559/tests/test_layers.py#L621](https://github.com/onnx/keras-onnx/blob/810cc948fb9c3b0a04345bc556678c03a9b19559/tests/test_layers.py#L621)

Until this is fixed, a workaround is to manually insert transpose ops in the onnx graph before and after softmax. Otherwise the output won't make sense.

**Comment by jiafatom** [comment]
Thanks for raising this. Can you provide a model for us to debug?
Is this a keras model or tf.keras model? If it is keras model, we only support until tf 2.0 backend, because after that it is recommended to use tf.keras model.

**Comment by dd1923** [comment]
Oops I thought I pasted the colab link but that was a wrong paste. Updated the original post with a colab link with the full code.

**Comment by jiafatom** [comment]
We just checked in the fix. Can you pull latest keras2onnx master and try again?

**Comment by dd1923** [comment]
Yup. Tested and it works. That was super fast fix. Thanks!
